### ISSUE TITLE
Rroblem with Style/HashSyntax cop in Ruby 3.1

### ISSUE THREAD

**Issue by tycooon** [rubocop/rubocop#10359]
When using ruby 3.1 hash syntax, I get the `Lint/UselessAssignment` cop warning.

## Expected behavior

No warnings.

## Actual behavior

```
1.rb:4:3: W: Lint/UselessAssignment: Useless assignment to variable - user.
  user = double :user, id: 42, token:
  ^^^^

1 file inspected, 1 offense detected
```

## Steps to reproduce the problem

```ruby
def some_method(token)
  user = double :user, id: 42, token:
  described_class.build(user)
end
```

Version with brackets works as expected:

```ruby
def some_method(token)
  user = double(:user, id: 42, token:)
  described_class.build(user)
end
```

## RuboCop version

```
1.24.1 (using Parser 3.1.0.0, rubocop-ast 1.15.1, running on ruby 3.1.0 arm64-darwin21)
```

## Update:
It turns out that Ruby interprets the first one as

```ruby
  user = double :user, id: 42, token: described_class.build(nil)
```

So technically this is not a false positive. However, this also means that auto-correction for Style/HashSyntax cop is not safe in this case. So probably this should be fixed somehow.

**Comment by koic** [comment]
In Ruby 3.1, the following code is equivalent. So, line breaked expression is the hash value.

```ruby
def some_method(token)
  user = double :user, id: 42, token:
  described_class.build(user)
end
```

```ruby
def some_method(token)
  user = double :user, id: 42, token: described_class.build(user)
end
```

This behavior is being considered in the following Ruby core issue.
https://bugs.ruby-lang.org/issues/18396

> So technically this is not a false positive. However, this also means that auto-correction for Style/HashSyntax cop is not safe in this case. So probably this should be fixed somehow.

This issue has been resolved by https://github.com/rubocop/rubocop/pull/10357. It will be included in the next release.

**Comment by tycooon** [comment]
I tested with version 1.25.0 and it looks like it still breaks the code with auto-correction even in safe mode. In provided example, it would replace 

```ruby
build :user, token: token
SomeClass.new(true) 
```

with

```ruby
build :user, token: token
SomeClass.new(true)
```

Which breaks the code. Note that if there is `SomeClass.new` without an argument, it doesn't report a warning.

Also, now it doesn't do auto-correction in a lot of safe cases even with `EnforcedShorthandSyntax: always` option. So I guess some more fixes are needed.

**Comment by koic** [comment]
@tycooon Your feedback is correct. I've opened #10363 to resolve the issues. Thank you!

**Comment by tycooon** [comment]
Great, thank you 👍

**Comment by jbedworth** [comment]
I'm having this problem today (using Ruby 3.3.4 and the linter that comes with Shopify's VSCode ruby language server plugin).  I have a lot of code where I'm required to pass in named attributes (e.g.
some_method(user: user, token: token, params: {some_hash: 'works ok'})
the helpful formatter in VSCode breaks my code on save by changing it to:
some_method(user:, token:, params{some_hash: 'works ok'}. I can't figure out how to keep the valid formatter benefit without having my code broken every time I save.  Assume this is due to a bug in the linter.

**Comment by jbedworth** [comment]
> I'm having this problem today (using Ruby 3.3.4 and the linter that comes with Shopify's VSCode ruby language server plugin). I have a lot of code where I'm required to pass in named attributes (e.g. some_method(user: user, token: token, params: {some_hash: 'works ok'}) the helpful formatter in VSCode breaks my code on save by changing it to: some_method(user:, token:, params{some_hash: 'works ok'}. I can't figure out how to keep the valid formatter benefit without having my code broken every time I save. Assume this is due to a bug in the linter.

Ok, I guess making the jump from Ruby 2.2.3 to 3.3.4 (with a 10 year gap) I missed some stuff.  Looks like
method(foo: foo) and method(foo:) are the same now.  Looks broken, but seems to work, so... ok.
